include/pybind11/stl.h

```diff
@@ -159,10 +159,12 @@ template <typename Type, typename Value> struct list_caster {
     }
 
 private:
-    template <typename T = Type,
-              enable_if_t<std::is_same<decltype(std::declval<T>().reserve(0)), void>::value, int> = 0>
-    void reserve_maybe(sequence s, Type *) { value.reserve(s.size()); }
-    void reserve_maybe(sequence, void *) { }
+    template <typename T = Type>
+    void reserve_maybe(decltype(std::declval<T>().reserve(0), iterable{}) it) {
+      if (hasattr(it, "__len__"))
+          value.reserve(len(it));
```

Comment:
`hasattr(it, "__len__")` is very inefficient, considering that we know that we're dealing with a sequence. For this use case, CPython provides `PyObject_LengthHint()` since python 3.4. To save you the trouble of error checking, pybind11 provides `len_hint()`.